File: python/valid-parentheses.py
"""
通过dict和stack判断是否是valid
"""

class Solution:
    # @return a boolean
    def isValid(self, s):
        stack = []
        dic = {'(': ')', '{': '}', '[': ']'}

        for i in s:
            if i in dic.keys():
                stack.append(i)
            elif i in dic.values():
                if len(stack) == 0 or i != dic[stack.pop()]:
                    return False
            else:
                return False

        return len(stack) == 0


# test
print Solution().isValid("({[]})")
print Solution().isValid("({[]}]")
print Solution().isValid("(")
print Solution().isValid(")")
print Solution().isValid("")

File: cpp/valid-parentheses.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        int length = s.size();
        if (length == 0) {
            return true;
        }

        vector<char> stack;
        for (char c : s) {
            if (c == '[' || c == '{' || c == '(') {
                stack.push_back(c);
            } else if (c == ']' && !stack.empty() && stack.back() == '[') {
                stack.pop_back();
            } else if (c == '}' && !stack.empty() && stack.back() == '{') {
                stack.pop_back();
            } else if (c == ')' && !stack.empty() && stack.back() == '(') {
                stack.pop_back();
            } else {
                return false;
            }
        }

        return stack.empty();
    }
};


int main () {
    vector<string> vs {"[", "]" ,"", "[{()}{[()][]}]", "[][](){}"};

    Solution s;
    for (string str : vs) {
        cout << s.isValid(str) << endl;
    }

    return 0;
}
